flink-staging/flink-ml/src/main/scala/org/apache/flink/ml/preprocessing/MinMaxScaler.scala

```diff
@@ -0,0 +1,254 @@
+    * @return
+    */
+  implicit def fitVectorMinMaxScaler[T <: Vector] = new FitOperation[MinMaxScaler, T] {
+    override def fit(instance: MinMaxScaler, fitParameters: ParameterMap, input: DataSet[T])
+    : Unit = {
+      val metrics = extractFeatureMinMaxVectors(input)
+
+      instance.metricsOption = Some(metrics)
+    }
+  }
+
+  /** Trains the [[MinMaxScaler]] by learning the minimum and maximum of the features of the
+    * training data which is of type [[LabeledVector]]. The minimum and maximum are used to
+    * transform the given input data.
+    *
+    */
+  implicit val fitLabeledVectorMinMaxScaler = {
+    new FitOperation[MinMaxScaler, LabeledVector] {
+      override def fit(
+        instance: MinMaxScaler,
+        fitParameters: ParameterMap,
+        input: DataSet[LabeledVector])
+      : Unit = {
+        val vectorDS = input.map(_.vector)
+        val metrics = extractFeatureMinMaxVectors(vectorDS)
+
+        instance.metricsOption = Some(metrics)
+      }
+    }
+  }
+
+  /** Calculates in one pass over the data the features' minimum and maximum values.
+    *
+    * @param dataSet The data set for which we want to calculate the minimum and maximum values.
+    * @return  DataSet containing a single tuple of two vectors (minVector, maxVector).
+    *          The first vector represents the minimum values vector and the second is the maximum
+    *          values vector.
+    */
+  private def extractFeatureMinMaxVectors[T <: Vector](dataSet: DataSet[T])
+  : DataSet[(linalg.Vector[Double], linalg.Vector[Double])] = {
+
+    val minMax = dataSet.map {
+      v => (v.asBreeze, v.asBreeze)
+    }.reduce {
+      (minMax1, minMax2) => {
+
+        val tempMinimum = min(minMax1._1, minMax2._1)
+        val tempMaximum = max(minMax1._2, minMax2._2)
+
+        (tempMinimum, tempMaximum)
+      }
+    }
+    minMax
+  }
+
+  /** [[TransformOperation]] which scales input data of subtype of [[Vector]] with respect to
+    * the calculated minimum and maximum of the training data. The minimum and maximum
+    * values of the resulting data is configurable.
+    *
+    * @tparam T Type of the input and output data which has to be a subtype of [[Vector]]
+    * @return
+    */
+  implicit def transformVectors[T <: Vector : BreezeVectorConverter : TypeInformation : ClassTag]
+  = {
+    new TransformOperation[MinMaxScaler, T, T] {
+      override def transform(
+        instance: MinMaxScaler,
+        transformParameters: ParameterMap,
+        input: DataSet[T])
+      : DataSet[T] = {
+
+        val resultingParameters = instance.parameters ++ transformParameters
+        val min = resultingParameters(Min)
+        val max = resultingParameters(Max)
+
+        instance.metricsOption match {
+          case Some(metrics) => {
+            input.mapWithBcVariable(metrics) {
+              (vector, metrics) => {
+                val (broadcastMin, broadcastMax) = metrics
+                var myVector = vector.asBreeze
```

Comment:
Can avoid some code duplication here by having a transformVector(vector: flink.ml.Vector, metrics: (bMin, bMax)): flink.ml.Vector function. That can then get called inside the map for both the Vector and LabedledVector case.